mockgcp/mockcompute/operations.go

```diff
@@ -0,0 +1,176 @@
+
+	op.Status = PtrTo(pb.Operation_DONE)
+
+	log.Info("storing operation", "fqn", fqn)
+	if err := s.storage.Create(ctx, fqn, op); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating LRO: %v", err)
+	}
+	return op, nil
+}
+
+func (s *computeOperations) startLRO0(ctx context.Context, op *pb.Operation, fqn string, callback func() (proto.Message, error)) (*pb.Operation, error) {
+	log := klog.FromContext(ctx)
+
+	now := time.Now()
+	nanos := now.UnixNano()
+
+	if op == nil {
+		op = &pb.Operation{}
+	}
+
+	op.StartTime = PtrTo(formatTime(now))
+	op.InsertTime = PtrTo(formatTime(now))
+	op.Id = PtrTo(uint64(nanos))
+
+	op.Progress = PtrTo(int32(0))
+
+	op.Kind = PtrTo("compute#operation")
+	op.SelfLink = PtrTo("https://compute.googleapis.com/compute/v1/" + fqn)
+
+	op.Status = PtrTo(pb.Operation_RUNNING)
+
+	log.Info("storing operation", "fqn", fqn)
+	if err := s.storage.Create(ctx, fqn, op); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating LRO: %v", err)
+	}
+
+	go func() {
+		result, err := callback()
+		finished := &pb.Operation{}
+		if err2 := s.storage.Get(ctx, fqn, finished); err2 != nil {
+			klog.Warningf("error getting LRO: %v", err2)
+			return
+		}
+
+		finished.Progress = PtrTo(int32(100))
+		finished.Status = PtrTo(pb.Operation_DONE)
+		finished.EndTime = PtrTo(formatTime(time.Now()))
+
+		if err != nil {
+			klog.Warningf("TODO: handle LRO error %v", err)
+		} else {
+			klog.Warningf("TODO: handle LRO result %v", result)
+		}
+		if err := s.storage.Update(ctx, fqn, finished); err != nil {
+			klog.Warningf("error updating LRO: %v", err)
+			return
+		}
+	}()
+
+	return op, nil
+}
+
+func (s *computeOperations) startRegionalLRO(ctx context.Context, projectID string, region string, op *pb.Operation, callback func() (proto.Message, error)) (*pb.Operation, error) {
+	now := time.Now()
+	millis := now.UnixMilli()
+	id := uuid.NewUUID()
+
+	name := fmt.Sprintf("operation-%d-%s", millis, id)
+	fqn := s.regionalOperationFQN(projectID, region, name)
+
+	op.Name = PtrTo(name)
+	return s.startLRO0(ctx, op, fqn, callback)
+}
+
+func (s *computeOperations) startGlobalLRO(ctx context.Context, projectID string, op *pb.Operation, callback func() (proto.Message, error)) (*pb.Operation, error) {
+	now := time.Now()
+	millis := now.UnixMilli()
+	id := uuid.NewUUID()
+
+	name := fmt.Sprintf("operation-%d-%s", millis, id)
+	fqn := s.globalOperationFQN(projectID, name)
+
+	op.Name = PtrTo(name)
+	return s.startLRO0(ctx, op, fqn, callback)
+}
+
+// Gets the latest state of a long-running operation.  Clients can use this
+// method to poll the operation result at intervals as recommended by the API
+// service.
+func (s *computeOperations) getOperation(ctx context.Context, fqn string) (*pb.Operation, error) {
+	op := &pb.Operation{}
+	if err := s.storage.Get(ctx, fqn, op); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("LRO not found for %v", fqn)
```

Comment:
nit: I do not think the info adds any value here, since we have the error message below

Reply:
Good point; removed